src/main/java/picard/illumina/ExtractIlluminaBarcodes.java

```diff
@@ -218,7 +213,9 @@
     private IlluminaDataProviderFactory factory;
 
     private final Map<String, BarcodeMetric> barcodeToMetrics = new LinkedHashMap<>();
-    private final ConcurrentHashMap<String, PerTileBarcodeExtractor.BarcodeMatch> barcodeLookupMap = new ConcurrentHashMap<>();
+
+    private final static int MAX_LOOKUP_SIZE = 100000;
```

Comment:
I'm curious to see if the timings I go for the MiSeq run make a bigger difference with a novaseq worth of data. Perhaps this should be a tweakable parameter?

```
390625 (5^8) = Runtime: 19.324 s
128k          = Runtime: 19.435 s
106496        = Runtime: 19.048 s
100k          = Runtime: 19.241 s
64k           = Runtime: 20.215 s
32k           = Runtime: 19.65 s
16k           = Runtime: 19.55 s
8k            = Runtime: 19.305 s
4k            = Runtime: 18.516 s
1             = Runtime: 29.713 s
```

Reply:
I tried a little bit with the MiSeq run and it seemed to me that expanding the cache further slowed things down.  My 2c is that the right solution (yet to be proven):
1. Extract/remove any reference to Metric objects from the _matching_ methods so that a) they're cleaner and b) we can pre-fill the cache without having to worry about messing up metrics
2. Pre-fill the cache with all perfect matches, and all no-calls
3. Change the conditional for entry into the cache to be based on whether the barcode read matched a known barcode and _not_ limited by number.

My thought being that there is a lot of value caching things that we see frequently, and little value caching things that we are unlikely to see frequently.

Reply:
The latest code includes the above changes.